### PR TITLE
allow dynamic (bound) location to firebase element

### PR DESCRIPTION
**Pull request by christophe-g**
Allow this type of custom element declaration : 
```html
<firebase-document id="base" location="{{location}}" data="{{data}}" log>
</firebase-document>
```
or
```html
<firebase-document id="base" location="{{userUrl}}/settings" data="{{data}}" log>
</firebase-document>
```
if this PR is accepted, should close https://github.com/GoogleWebComponents/firebase-element/issues/35

C.

### CODE REVIEW

File: firebase-collection.html [hunk 1]
```diff
@@ -384,37 +393,50 @@
 
         this._log('Recomputing query.', arguments);
 
-        query = new Firebase(location);
+        try {
+          query =  new Firebase(location);
+          
+          this.queryString = '';
```
Comment (MeTaNoV) [inline firebase-collection.html]: queryError instead?

File: firebase-collection.html [hunk 2]
```diff
@@ -384,37 +393,50 @@
...
+
+          if (limitToFirst != null) {
+            query = query.limitToFirst(limitToFirst);
+          }
 
-        if (limitToFirst != null) {
-          query = query.limitToFirst(limitToFirst);
+        } catch(e) {
+          
+          this.queryString = e.toString();
```
Comment (MeTaNoV) [inline firebase-collection.html]: queryError instead?

File: firebase-document.html [hunk 3]
```diff
@@ -90,24 +90,52 @@
...
-      return new Firebase(location);
+      var query;
+      try {
+      
+        query =  new Firebase(location);
+        this.queryString = '';
+      
+      } catch(e) {
+        
+        this.queryString = e.toString();
```
Comment (MeTaNoV) [inline firebase-document.html]: queryError instead?

File: firebase-document.html [hunk 4]
```diff
@@ -90,24 +90,52 @@
       if (!location) {
         return;
       }
-
-      return new Firebase(location);
+      var query;
+      try {
+      
+        query =  new Firebase(location);
+        this.queryString = '';
```
Comment (MeTaNoV) [inline firebase-document.html]: queryError instead?

**Comment by MeTaNoV** [comment]
did you test the second case? The Polymer team will also probably ask you for some test! Be prepared! :)

**Comment by christophe-g** [comment]
@MeTaNoV Thanks for pointing out !

Yes, second case is working for me - will add a test, that's a good idea.
Cheers,
C.

**Comment by MeTaNoV** [comment]
@christophe-g Pas de soucis! :) And if you are working with Firebase, this might interest you: https://github.com/MeTaNoV/firebase-element-extended

**Comment by christophe-g** [comment]
@MeTaNoV 
Danke vielmals. I did see your extension... after I implemented my own elements for auth and sign-in. They are still a bit rough, so might come back to your work. 
C.

File: firebase-collection.html [hunk 5]
```diff
@@ -253,6 +253,10 @@
 
         this._log('Adding new item to collection with value:', data);
 
+        if(!this.query) {
```
Comment (ebidel) [inline firebase-collection.html]: spacing on these conditionals `if (`

File: firebase-collection.html [hunk 6]
```diff
@@ -384,37 +393,50 @@
 
         this._log('Recomputing query.', arguments);
 
-        query = new Firebase(location);
+        try {
+          query =  new Firebase(location);
```
Comment (ebidel) [inline firebase-collection.html]: remove 1 space between = and new

Comment (ebidel) [inline firebase-collection.html]: is there a reason you've wrapped this in try/catch. not sure we really need that. There's probably a better way.

File: firebase-query-behavior.html [hunk 7]
```diff
@@ -33,6 +33,16 @@
         readOnly: true
       },
 
+     /**
+       * A string storing error when connection to firebase instance is not successfull (for instance when location is a bound property - not available when instanciating the connection)
```
Comment (ebidel) [inline firebase-query-behavior.html]: instantiating

Comment (ebidel) [inline firebase-query-behavior.html]: successful

File: firebase-document.html [hunk 8]
```diff
@@ -90,24 +90,52 @@
...
+      var query;
+      try {
+      
+        query =  new Firebase(location);
+        this.queryError = '';
+      
+      } catch(e) {
+        
+        this.queryError = e.toString();
+        this._error('Query cannot be instanciated with location ' + location + ' (' + this.queryError + ')');
```
Comment (ebidel) [inline firebase-document.html]: instantiated

File: firebase-collection.html [hunk 9]
```diff
@@ -384,37 +393,50 @@
...
+          if (limitToFirst != null) {
+            query = query.limitToFirst(limitToFirst);
+          }
 
-        if (limitToFirst != null) {
-          query = query.limitToFirst(limitToFirst);
+        } catch(e) {
+          
+          this.queryError = e.toString();
+          this._error('Query cannot be instanciated with location ' + location + ' (' + this.queryError + ')');
```
Comment (ebidel) [inline firebase-collection.html]: instantiated

File: firebase-collection.html (hunk 6 above)
Comment (christophe-g) [inline firebase-collection.html]: the reason is that the dynamic location might not be valid in some use case. For instance `location="{{userPath}}/settings" ` will resolve to `settings` when user is not yet logged in. Trying to access firebase will result in an error in this case.
C.

**Comment by chuckh** [comment]
LGTM

**Comment by addyosmani** [comment]
LGTM2. Will wait on @ebidel to make a final merge call

File: firebase-collection.html [hunk 10]
```diff
@@ -326,6 +330,11 @@
       },
 
       _applySubPathChange: function(change) {
+
+        if(!this.query) {
```
Comment (ebidel) [inline firebase-collection.html]: if (!this.query) {

File: firebase-collection.html [hunk 11]
```diff
@@ -384,37 +393,50 @@
...
 
-        if (limitToFirst != null) {
-          query = query.limitToFirst(limitToFirst);
+          if (limitToFirst != null) {
+            query = query.limitToFirst(limitToFirst);
+          }
+
+        } catch(e) {
+          
+          this.queryError = e.toString();
```
Comment (ebidel) [inline firebase-collection.html]: Shoudl queryError should be public? Probably sould be private: `_queryError`

File: firebase-query-behavior.html [hunk 12]
```diff
@@ -33,6 +33,16 @@
...
       },
 
+     /**
+       * A string storing error when connection to firebase instance is not successfull (for instance when location is a bound property - not available when instantiating the connection)
+       * 
+       */
+      queryError: {
+        type: String,
+        value: '',
+        notify: true
```
Comment (ebidel) [inline firebase-query-behavior.html]: I think instead of adding a notifying property, it would be better to just fire an event that can be caught?

Comment (christophe-g) [inline firebase-query-behavior.html]: Agreed, an event notifying the error is more appropriate in the case. Will update the PR. 
Thanks for the review ; ) 
C.

**Comment by ebidel** [comment]
LGTM. just one more nit

**Comment by chuckh** [comment]
@christophe-g  please change `this.fire('query-error', error);` to `this.fire('firebase-query-error', error);` at line: https://github.com/GoogleWebComponents/firebase-element/pull/109/files#diff-aa262c2b55d33a1a972f74c57338f065R171.  Thanks.

**Comment by christophe-g** [comment]
@chuckh 
Done.
C.

**Comment by chuckh** [comment]
LGTM. Thanks @christophe-g!  @ebidel what do you think?

**Comment by ebidel** [comment]
:+1:

**Pull request merged**